### PR TITLE
feat: Google API Auth

### PR DESCRIPTION
**Pull request by crflanigan**
### Required for all PRs:

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

resolves  https://github.com/influxdata/telegraf/issues/7913

Enhanced the HTTP output plugin enabling it to authenticate against a Google API protected by the OAuth 2.0 protocol.

### CODE REVIEW

**Comment by reimda** [comment]
Looks like this is a continuation of #10369, right @crflanigan?

**Comment by crflanigan** [comment]
> Looks like this is a continuation of #10369, right @crflanigan?

Sure is. We had some git issues that we resolved and since the old PR description had drifted based on feedback from the community we went ahead and created a updated one.

**Comment by crflanigan** [comment]
@srebhan @Hipska 
Here's the new PR.

**Review by Hipska — APPROVED** [review]
Looks all good to me, however I would add some extra info in the README file on how to get that credentials file or how it does look like.

File: plugins/common/http/config.go
Comment (srebhan) [inline plugins/common/http/config.go:14]: Can you please order the import by

1. built-in libraries such as `fmt`, `net/http`, `time`, etc
2. 3rd party libraries e.g. everything starting with ` github.com/...` except for telegraf internal ones
3. telegraf internal libraries, i.e. everything starting with `github.com/influxdata/telegraf`

Separate all those groups by a newline and sort alphabetically within each group.

File: plugins/common/oauth/config.go
Comment (srebhan) [inline plugins/common/oauth/config.go:8]: Please move up again, see my comment above.

File: plugins/outputs/http/http_test.go
Comment (srebhan) [inline plugins/outputs/http/http_test.go:543]: Is it required to set the port here? The reason I'm asking is that if two tests run in parallel one might fail due to the port being already taken...

File: plugins/outputs/http/http_test.go
Comment (srebhan) [inline plugins/outputs/http/http_test.go:551]: Make this a const?

File: plugins/outputs/http/http_test.go
Comment (srebhan) [inline plugins/outputs/http/http_test.go:608]: Optionally you can shorten this to
```suggestion
			tt.plugin.SetSerializer(influx.NewSerializer())
			require.NoError(t, tt.plugin.Connect())
			require.NoError(t, tt.plugin.Write([]telegraf.Metric{getMetric()}))
```
but it's not mandatory. If you like to keep it as is, please just mark as resolved.

**Review by srebhan — COMMENTED** [review]
Wow that looks very clean @crflanigan and @zachmares! I just have some small comments regarding import ordering and one regarding choosing a fixed port. Can you please take a look?

File: plugins/common/http/config.go
Comment (zachmares) [reply plugins/common/http/config.go]: I found a handy setting in Goland for this request. Reordered!

File: plugins/common/oauth/config.go
Comment (zachmares) [reply plugins/common/oauth/config.go]: Also reordered.

File: plugins/outputs/http/http_test.go
Comment (zachmares) [reply plugins/outputs/http/http_test.go]: Hey, always down to reduce a few lines in code as long as it isn't confusing. 😎

File: plugins/outputs/http/http_test.go
Comment (zachmares) [reply plugins/outputs/http/http_test.go]: Done.

File: plugins/outputs/http/http_test.go
Comment (zachmares) [reply plugins/outputs/http/http_test.go]: The original reason I set a static port is because the `token_uri` in the key file is used to request the access token. Since I couldn't update the port of the token_uri inside the key file, I pinned it. 

I have since refactored this test to use `func (c *common) TempDir() string` and `func CreateTemp(dir string, pattern string) (*File, error)`, which obviates the need for the key file in the testdata folder.

**Review by Hipska — APPROVED** [review]

File: plugins/outputs/http/http.go
Comment (crflanigan) [reply plugins/outputs/http/http.go:176]: Updated.

Comment (crflanigan) [reply plugins/outputs/http/http.go:176]: @Hipska

Comment (Hipska) [reply plugins/outputs/http/http.go:176]: Yeah, but only in one error case. Why no message for the other case?

**Review by Hipska — APPROVED** [review]

File: plugins/outputs/http/http.go
Comment (srebhan) [inline plugins/outputs/http/http.go:181]: Maybe just one more suggestion... If you move the token validity check into `GetAccessToken` and make that function return the token, you can mimic the other code around this function, i.e.
```suggestion
	if h.HTTPClientConfig.OAuth2Config.CredentialsFile != "" {
		token, err := h.HTTPClientConfig.OAuth2Config.GetAccessToken(context.Background(), h.URL)
		if err != nil {
			return err
		}
		token.SetAuthHeader(req)
	}
```

It would hide the whole refresh logic in `h.HTTPClientConfig.OAuth2Config` and you can also "unexport" `AccessToken` avoiding direct manipulation.

What do you think? I would not hold up the whole thing just for that, but it looks cleaner to me...

Comment (zachmares) [reply plugins/outputs/http/http.go:181]: Once again, great suggestion!

**Review by srebhan — APPROVED** [review]
Looks good to me. Well done sons of awesome! ;-) Thanks for the great work @crflanigan and @zachmares!

**Comment by crflanigan** [comment]
> Looks good to me. Well done sons of awesome! ;-) Thanks for the great work @crflanigan and @zachmares!

Thank you for the support!

File: plugins/outputs/http/http.go
Comment (reimda) [inline plugins/outputs/http/http.go:178]: HTTPClientConfig is also used by plugins/inputs/http and plugins/inputs/openstack. Would it make sense to move this section to a function and call it in those two plugins too?

File: plugins/outputs/http/README.md
Comment (reimda) [inline plugins/outputs/http/README.md:115]:
```suggestion
The `credentials_file` setting is used with Google Cloud APIs. It specifies the json key file. To learn about creating Google service accounts, consult Google's
[oauth2 service account documentation][create_service_account]. An example use case is a metrics proxy deployed to
Cloud Run. In this example, the service account must have the "run.routes.invoke" permission.
```

File: plugins/outputs/http/README.md
Comment (reimda) [inline plugins/outputs/http/README.md:32]: The name `credentials_file` is very generic. If this is a Google Cloud specific feature, could we rename the setting `gcp_credentials_file` or something else that indicates it's only for google apis?

Comment (zachmares) [reply plugins/outputs/http/README.md:32]: This is indeed Google-specific, so yes `gcp_credentials_file` works.

Comment (zachmares) [reply plugins/outputs/http/README.md:32]: Or perhaps `google_application_credentials`. Please let me know your preference.

Comment (reimda) [reply plugins/outputs/http/README.md:32]: `google_application_credentials` sounds good to me. It's a little long but it's descriptive and it's the same as the environment variable the cli tools use, just lowercase to fit in toml, so it should be familiar for gcp users.

https://cloud.google.com/docs/authentication/getting-started#setting_the_environment_variable

Comment (zachmares) [reply plugins/outputs/http/README.md]: Done!

File: plugins/outputs/http/http.go
Comment (reimda) [reply plugins/outputs/http/http.go:178]: @zachmares what do you think about adding it to these other plugins? I don't really like the idea of adding an undocumented variable to other plugins if it won't ever be useful.

Comment (zachmares) [reply plugins/outputs/http/http.go:178]: @reimda Apologies, I missed this comment yesterday. I'm not entirely sure at the moment. I'd need to defer to the experts for this one.

Comment (crflanigan) [reply plugins/outputs/http/http.go:178]: Hi @reimda @srebhan, 
This PR was really just for getting through the Google auth using the HTTP output plugin. 
I can see that it could make sense to make this available to other plugins, but being that this doesn't fall within the scope of the description or the issue it's trying to solve for I would say it should be a seperate PR.
Does that make sense?

File: plugins/outputs/http/README.md
Comment (crflanigan) [reply plugins/outputs/http/README.md:115]: @zachmares Did we address this?

**Comment by Hipska** [comment]
All tests seem to succeed..

File: plugins/outputs/http/README.md
Comment (zachmares) [reply plugins/outputs/http/README.md]: Committed the suggested changes. Thank you @reimda for the word-smithing!

File: plugins/outputs/http/http.go
Comment (zachmares) [reply plugins/outputs/http/http.go:178]: I'm not sure oauth makes sense in the context of a Telegraf input plugin. This contribution sets auth headers on the payloads written from the output plugin.

I'm happy to discuss and explore this further, but I agree with @crflanigan that those discussions are better suited for a future and separate PR. A use case would be helpful too. 

I believe this contribution is self-contained and ready to go.

**Review by Hipska — APPROVED** [review]

**Comment by zachmares** [comment]
"plugins/outputs/graylog TestWriteTCP" passes when I run `CGO_ENABLED=1 make test` from my development environment. Re-run test? Requesting some intervention on this one... @srebhan

File: plugins/outputs/http/http.go
Comment (reimda) [reply plugins/outputs/http/http.go:178]: I understand that the scope of this pr is really only the http output, but adding the setting to the existing OAuth2Config struct has the side effect of also adding it to the http and openstack inputs. That makes the PR not ready to go.

Having CredentialsFile in inputs/http and inputs/openstack might not seem important, but are plans to add an api to query and change settings: #7519. That will expose the setting to users.

We could ignore the problem now, but if it turns out we need to remove it from OAuth2Config in the future because it doesn't apply to the inputs, it will not be possible to leave it in for the outputs. We want to plan for compatibility for the http output plugin and not add a setting we're likely to remove. It's better to find the right place for the setting now than rush this PR and get it wrong.

If it doesn't makes sense to add CredentialsFile to the http input and the openstack input, it should not be in OAuth2Config. If it is only going to be useful in the http output plugin, it should be in plugins/outputs/http/http.go's HTTP struct, not a shared struct like OAuth2Config or HTTPClientConfig.

**Review by reimda — CHANGES_REQUESTED** [review]
Marking "request changes" to show we're not ready to merge

File: plugins/outputs/http/http.go
Comment (Hipska) [reply plugins/outputs/http/http.go:178]: I can imagine there are APIs you want to pull with the http input that require Oauth2. I do agree not to implement it in input plugin for now as there is no need for it (yet?)

Comment (zachmares) [reply plugins/outputs/http/http.go]: I moved the Google API Auth settings and code to outputs/http.go as requested.

@reimda @srebhan Please take a look when you get the chance.

Comment (srebhan) [reply plugins/outputs/http/http.go]: After a short discussion on Slack, @zachmares agrees that this might also be helpful to at least `inputs.http`, so I suggest to keep the code in `common` and add the 
```golang
	// google api auth
	if h.HTTPClientConfig.OAuth2Config.CredentialsFile != "" {
		token, err := h.HTTPClientConfig.OAuth2Config.GetAccessToken(context.Background(), h.URL)
		if err != nil {
			return err
		}
		token.SetAuthHeader(req)
	}
```
part to both `http` input and output plugin at least.

Comment (srebhan) [reply plugins/outputs/http/http.go]: Sorry, seems like I got it wrong. @zachmares, can you please state if this auth method is useful for anything beyond `outputs.http` or not!? If it is *only* useful for `outputs.http` it should move there, otherwise, please keep it in common and add the above code to any plugin where it makes sense...

Sorry for the confusion I caused!

Comment (zachmares) [reply plugins/outputs/http/http.go]: No worries, Sven! As far as @crflanigan and I know, this contribution only makes sense in the outputs/http context. To recap, our use case is using Telegraf to gather system metrics and flushing them to an endpoint hosted in GCP where only authenticated invocations of the service are allowed.

If a use case arises wherein the an inputs plugin wishes to invoke a service in GCP, we'd be happy to move the code back to common.

Comment (zachmares) [reply plugins/outputs/http/http.go]: Although there are other output plugins that could benefit from this, such as `/plugins/outputs/bigquery` and `plugins/outputs/cloud_pubsub`...

**Review by reimda — APPROVED** [review]

**Pull request merged**
